### PR TITLE
Use neat periodic tasks for updating the preview versions.

### PR DESCRIPTION
**Pull request by isoos**
#4404

### CODE REVIEW

**Comment by mit-mit** [comment]
@isoos can you describe what experience this would give us?

**Comment by jonasfj** [comment]
We dynamically fetch information from `dart.dev` to detect what the latest stable version is.
We then periodically scan all packages and adjust "stable", "preview" and "prereleases" as the designated versions might change when a new version is published.

This doesn't happen with traffic switching because we don't store the "stable", "preview" and "prereleases" designations on a per-deployment basis -- that would probably introduce more latency.

Publishing a new stable release on `dart.dev` (in GCS), will start update "stable", "preview" and "prereleases" designations on pub.dev

**Comment by mit-mit** [comment]
Ah, that makes sense. We usually publish on dart.dev/in GCS some hours, or even days, in advance of announcing.

What is the time span after publishing the new stable release before pub.dev will have completed the update?

File: app/test/tool/neat_task/datastore_status_provider_test.dart [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    testWithProfile('set status', fn: () async {
+      final p1 = DatastoreStatusProvider.create(dbService, 'task-id');
+      expect(await p1.set([1, 2]), isTrue);
+      expect(await p1.get(), [1, 2]);
+
+      final p2 = DatastoreStatusProvider.create(dbService, 'task-id');
+      expect(await p2.set([3, 4]), isFalse);
+      expect(await p2.get(), [1, 2]);
+      expect(await p2.set([3, 4]), isTrue);
+      expect(await p2.get(), [3, 4]);
```
Comment (jonasfj) [inline app/test/tool/neat_task/datastore_status_provider_test.dart:37]:
```suggestion
      expect(await p2.get(), [3, 4]);
      
      expect(await p1.set([1, 2]), isFalse);
```

**Review by jonasfj — APPROVED** [review]

**Comment by isoos** [comment]
Currently we query the https://storage.googleapis.com/dart-archive/channels/stable/release/latest/VERSION endpoint. Not sure when this is being updated in the release process, but we query it every hour and may update the `/versions` list on the package page based on that. This PR adds a daily background task (that may run more frequently) to update individual Package entries in datastore. In an unlucky timing it may take 25 hours to get them updated.

**Comment by mit-mit** [comment]
Would it be possible to manually "kick" the background task to make it run quicker?

**Comment by jonasfj** [comment]
> Ah, that makes sense. We usually publish on dart.dev/in GCS some hours, or even days, in advance of announcing.

@athomas, when do we update the `https://storage.googleapis.com/dart-archive/channels/stable/release/latest/VERSION` file?

> What is the time span after publishing the new stable release before pub.dev will have completed the update?

We run the periodic task every 24 hours.

**Comment by athomas** [comment]
> @athomas, when do we update the `https://storage.googleapis.com/dart-archive/channels/stable/release/latest/VERSION` file?
 
https://storage.googleapis.com/dart-archive/channels/stable/release/latest/VERSION is updated just before the release goes to dart.dev.

It's copied from:
https://storage.googleapis.com/dart-archive/channels/stable/raw/latest/VERSION (not exactly, because it's actually copied from the folder with the version number rather than `latest` but the `latest` URL is probably what you'd want to query).

Now, raw will also contain broken builds (assuming the build worked, but perhaps the tests didn't work, or the build failed on some platforms but succeeded on others), so it can give you a heads up but you don't want to just deploy raw to prod.

**Comment by jonasfj** [comment]
> https://storage.googleapis.com/dart-archive/channels/stable/release/latest/VERSION is updated just before the release goes to dart.dev.


Thanks, that's perfect!

**Pull request merged**